src/modalities/models/model_factory.py

```diff
@@ -103,7 +202,26 @@ def get_weight_initalized_model(model: nn.Module, model_initializer: ModelInitia
         Returns:
             nn.Module: The initialized model.
         """
-        model_initializer.initialize_in_place(model)
+
+        def reset_parameters_if_function_exists(module: nn.Module):
+            # Recursively apply to all submodules
+            for submodule in module.children():
+                reset_parameters_if_function_exists(submodule)
+
+            # Check if the module has the `reset_parameters` method
+            if hasattr(module, "reset_parameters") and callable(getattr(module, "reset_parameters")):
+                module.reset_parameters()
+
+        if ModelFactory._is_model_on_meta_device(model=model):
+            # Allocate buffers and sharded parameters on GPU
+            model = model.to_empty(device="cuda")
```

Comment:
Is my understanding correct that the model is a FSDP2 wrapped model since it is assumed that the model is already sharded?

Reply:
yes, see our documentation here:
[In the case of FSDP2, we first instantiate the model, wrap it with FSDP2 and then initialize the weights following a pre-defined distribution. The model_raw can be instantiated on a meta device. After sharding with FSDP2 (wrapped_model), the model is still on a meta device and only during the initialisation, the model is moved to the data device (see [here](https://github.com/Modalities/modalities/blob/a7f683a6202209dcf8e78cbff4ca68991e89e081/src/modalities/models/model_factory.py)).](https://github.com/Modalities/modalities/blob/b915fd1e6f2bdb5044b5d71dd455b5518accaf21/docs/fsdp1_vs_fsdp_2.md#L27-L28)